Added functionality for BoundaryNames in SideSetsFromBoundingBoxGenerator

socratesgorilla opened the pull request:
Closes #22230

Commit:
Added functionality for boundarynames in SideSetsFromBoundingBoxGenerator #22230

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1191984/) on 62c31de wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22293/site)

This comment will be updated on new commits.

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -39,9 +39,10 @@ SideSetsFromBoundingBoxGenerator::validParams()
+      "The list of boundaries on the specified block within the bounding box to be modified");
+  params.addRequiredParam<BoundaryName>(
+      "boundary_new", "Boundary on specified block within the bounding box to assign");
   params.addParam<bool>("boundary_id_overlap",
```
you have to keep the old parameters around, just deprecate them.

Then in the body constructor you have logic to handle the user passing deprecated parameters

socratesgorilla replied:
No downstream apps appeared to have been using them. If this is the case, surely the parameters could and should just be replaced.

GiudGiud replied:
we dont know if any user was relying on this. Their input files would not be attached to an app. We pretty much have to do this because we claim to be doing soft deprecation

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -95,14 +113,12 @@ SideSetsFromBoundingBoxGenerator::generate()
           // loop over provided boundary vector to check all side sets for all boundary ids
-          for (unsigned int boundary_id_number = 0; boundary_id_number < _boundary_id_old.size();
-               boundary_id_number++)
+          for (auto boundary_id : boundary_ids)
```
nice

GiudGiud commented on test/tests/mesh_modifiers/modifier_depend_order/depend_with_force_prepare.i:
```diff
@@ -19,20 +19,11 @@
     block_id = 1
     bottom_left = '-200 -1 -400'
     top_right = '200 1 0'
-    force_prepare = true
```
no this needs to stay this is the point of this test

socratesgorilla replied:
This was an accidental commit, I will remove the change, as I'm trying to leave mesh_modifier tests untouched due to their imminent deletion.

socratesgorilla commented:
These will fail due to unedited mesh_modifier tests and should be reran once that PR has been dealt with.

Commit:
Deprecated params instead of removing

Applied feedback from review

clang format

Removed remaining unused tests, fixed getting all boundary ids

Commit:
Added golds and new test file

Changed arg in test parameter

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1192657/) on 2de9470 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>549eee</th>
      <th colspan=4>#22293 2de947</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.29%</td>
      <td>84.29%</td>
      <td>+0.00%</td>
      <td>86.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80089</td>
      <td>80103</td>
      <td>+14</td>
      <td>26</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14928</td>
      <td>14928</td>
      <td>-</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22293/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22293/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22293/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 86.67% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -57,8 +66,12 @@ SideSetsFromBoundingBoxGenerator::SideSetsFromBoundingBoxGenerator(
     _block_id(parameters.get<SubdomainID>("block_id")),
-    _boundary_id_old(parameters.get<std::vector<BoundaryName>>("boundary_id_old")),
-    _boundary_id_new(parameters.get<boundary_id_type>("boundary_id_new")),
+    _boundaries_old(parameters.isParamSetByUser("boundary_id_old")
+                        ? parameters.get<std::vector<BoundaryName>>("boundary_id_old")
+                        : parameters.get<std::vector<BoundaryName>>("boundaries_old")),
```
this will error weirdly
you should have a default value for boundaries_old (for now, it s temporary)

then in the body of the constructor, check that one (at least and at most) of these two parameters is passed

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -57,8 +66,12 @@ SideSetsFromBoundingBoxGenerator::SideSetsFromBoundingBoxGenerator(
+                        : parameters.get<std::vector<BoundaryName>>("boundaries_old")),
+    _boundary_new(parameters.isParamSetByUser("boundary_id_new")
+                      ? parameters.get<BoundaryName>("boundary_id_new")
+                      : parameters.get<BoundaryName>("boundary_new")),
```
same

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -80,6 +93,23 @@ SideSetsFromBoundingBoxGenerator::generate()
+  // If the new boundary id/name is not valid, make it instead
+  if (boundary_id_new == Moose::INVALID_BOUNDARY_ID)
+    boundary_id_new =
+        boundary_ids.empty() ? 0 : cast_int<boundary_id_type>(*boundary_ids.rbegin() + 1);
```
why would boundary_ids be empty? Isnt it required?

also I dont get the logic of "last_value + 1". I think we need a whole new number if it s not valid. 
You should look at the MeshUtils for how it s done for subdomains (looks at all subdomain ids in the mesh, and takes the biggest + 1)

socratesgorilla replied:
Removing the empty check in next push. On the other thing, I don't understand what is confusing about it. If the boundary id number given for the name specified is invalid, it finds the last used boundary ID, and makes a new one with an ID right after, i.e. if the last boundary id is 2, it makes a boundary id of 3.

GiudGiud replied:
it s not the last boundary ID in all the mesh. 
It s the last in your local context. You could be using a boundary ID that is used elsewhere

GiudGiud replied:
will do
this conditional with a 0 is still an issue imo

GiudGiud replied:
boundary_ids is only for the names you passed in, it s not for the entire mesh

socratesgorilla replied:
```std::vector<libMesh::boundary_id_type>
getBoundaryIDs(const libMesh::MeshBase & mesh,
               const std::vector<BoundaryName> & boundary_name,
               bool generate_unknown)
{
  const libMesh::BoundaryInfo & boundary_info = mesh.get_boundary_info();
  const std::map<libMesh::boundary_id_type, std::string> & sideset_map =
      boundary_info.get_sideset_name_map();
  const std::map<libMesh::boundary_id_type, std::string> & nodeset_map =
      boundary_info.get_nodeset_name_map();

  std::set<libMesh::boundary_id_type> boundary_ids = boundary_info.get_boundary_ids();
```
?

socratesgorilla replied:
maybe it does, I can't really tell, I will just change

GiudGiud replied:
you passed a bunch of boundary names as an argument. You got the boundary ids for them. Not all boundary ids for all the mesh, that list got downsampled

GiudGiud commented on test/tests/meshgenerators/sidesets_bounding_box_generator/deprecated_param.i:
```diff
@@ -0,0 +1,24 @@
+[Mesh]
```
no need to test deprecated parameters btw. We ve kind of made that call to gain time.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1197690/) on e22746c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22293/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22293/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1cd91e61ff5b6c9d14ef83a94b6ae512e8a3f546`

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -57,12 +66,26 @@ SideSetsFromBoundingBoxGenerator::SideSetsFromBoundingBoxGenerator(
     _block_id(parameters.get<SubdomainID>("block_id")),
-    _boundary_id_old(parameters.get<std::vector<BoundaryName>>("boundary_id_old")),
-    _boundary_id_new(parameters.get<boundary_id_type>("boundary_id_new")),
+    _boundaries_old(),
+    _boundary_new(),
```
```suggestion
```
not really needed though not incorrect

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -57,12 +66,26 @@ SideSetsFromBoundingBoxGenerator::SideSetsFromBoundingBoxGenerator(
+    _boundaries_old = parameters.get<std::vector<BoundaryName>>("boundary_id_old");
+  else
+    mooseError(
+        "Either boundaries_old or boundary_id_old(deprecated) must be specified in the test.");
```
```suggestion
        "Either boundaries_old or boundary_id_old (deprecated) must be specified");
```

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -57,12 +66,26 @@ SideSetsFromBoundingBoxGenerator::SideSetsFromBoundingBoxGenerator(
+  else if (parameters.isParamSetByUser("boundary_id_new"))
+    _boundary_new = parameters.get<BoundaryName>("boundary_id_new");
+  else
+    mooseError("Either boundary_new or boundary_id_new(deprecated) must be specified in the test.");
```
```suggestion
    mooseError("Either boundary_new or boundary_id_new (deprecated) must be specified.");
```

GiudGiud commented on framework/src/meshgenerators/SideSetsFromBoundingBoxGenerator.C:
```diff
@@ -80,6 +103,22 @@ SideSetsFromBoundingBoxGenerator::generate()
+
+  // If the new boundary id/name is not valid, make it instead
+  if (boundary_id_new == Moose::INVALID_BOUNDARY_ID)
+    boundary_id_new = cast_int<boundary_id_type>(*(boundary_info.get_boundary_ids().rbegin()) + 1);
```
this is the right code, but we would want a utility for this because it s tricky
could you please add to mooseMeshUtils the equivalent of this:
```
/**
 * Checks input mesh and returns max(block ID) + 1, which represents
 * a block ID in the mesh that is not currently in use
 * @param input mesh over which to compute the next free block id
 */
SubdomainID getNextFreeSubdomainID(MeshBase & input_mesh);

```
for BoundaryIDs

GiudGiud reviewed (changes requested):
looks good, just these tiny changes and needs a rebase then good to go

Commit:
Abstracted get next boundary id to a moosemeshutils call

GiudGiud commented on framework/include/utils/MooseMeshUtils.h:
```diff
@@ -173,6 +173,13 @@ bool isCoPlanar(const std::vector<Point> vec_pts);
 SubdomainID getNextFreeSubdomainID(MeshBase & input_mesh);
 
+/**
+ * Checks input mesh and returns boundary_ids.rbegin() + 1, which represents
```
```suggestion
 * Checks input mesh and returns the largest boundary ID in the mesh plus one, which is
```

GiudGiud commented on framework/include/utils/MooseMeshUtils.h:
```diff
@@ -173,6 +173,13 @@ bool isCoPlanar(const std::vector<Point> vec_pts);
 
+/**
+ * Checks input mesh and returns boundary_ids.rbegin() + 1, which represents
+ * a boundary ID in the mesh that is not currently in use
```
```suggestion
 * a boundary ID that is not currently in use in the mesh
```

GiudGiud commented on framework/src/utils/MooseMeshUtils.C:
```diff
@@ -321,6 +321,14 @@ getNextFreeSubdomainID(MeshBase & input_mesh)
+    return 0;
+  return (*boundary_ids.rbegin() + 1);
+}
 bool
```
```suggestion

bool
```

GiudGiud commented on framework/src/utils/MooseMeshUtils.C:
```diff
@@ -321,6 +321,14 @@ getNextFreeSubdomainID(MeshBase & input_mesh)
   }
 }
 
+boundary_id_type
```
```suggestion
BoundaryID
```

GiudGiud commented on test/tests/mesh_modifiers/modifier_depend_order/depend_with_force_prepare.i:
```diff
@@ -0,0 +1,30 @@
+[Mesh]
```
these cant come back

GiudGiud commented on test/tests/transfers/multiapp_reporter_transfer/main.i:
```diff
@@ -46,6 +46,7 @@
     real_vector_values = '0'
     string_names = str
     string_values = 'foo'
+    execute_on = 'timestep_end'
```
```suggestion
```

GiudGiud commented on test/tests/transfers/multiapp_reporter_transfer/main.i:
```diff
@@ -94,6 +95,7 @@
     from_reporters = 'from_main_rep/num'
     to_multi_app = sub
     subapp_index = 0
+    execute_on = 'timestep_begin'
```
```suggestion
```
